### ISSUE TITLE
Use XLA operators instead of custom threefry2x32 kernel to generate random number, due to high memory costs of threefry2x32 kernel.

### ISSUE THREAD

**Issue by shawnwang18** [google/jax#20328]
### Description

It seems that Jax's default random number generation algorithm threefry2x32 consumes large amout of memory.

From search results, the requirement of threefry2x32 is:

To generate random values using threefry2x32 directly, you need to provide a counter array and a PRNG key. The counter array determines the shape of the output, while the PRNG key determines the specific random values generated

The counter and key are arrays of unsigned 32-bit integers. For threefry2x32, both the counter and key have 2 elements

This means to generate a random tensor, custom kernel threefry2x32 requires 2X memory size of the output tensor. This will result in big issue about memory consumption, especially for embedding tensor, as embedding tensor is already very large, 2X memory consumption is very likely to cause OOM.

An example of HLO dumps is attached. 
[module_0001.jit_init.sm_9.0_gpu_after_optimizations.txt](https://github.com/google/jax/files/14661250/module_0001.jit_init.sm_9.0_gpu_after_optimizations.txt)

Another solution is  to comment out this line of code in JAX:
https://github.com/google/jax/blob/b6e985ffe7af6054a289613eb70ce31c6360cd76/jax/_src/prng.py#L1045

If you comment it out, JAX will use an HLO-only implementation of the prng, with no CUDA. The only reason the custom kernel exists is because of compile time, and I'm not even sure it's needed any more. The HLO-only implementation can be fused with producers and consumers, so XLA may do something better in that case. 

For our reproduce code, the initialization module’s memory consumption reduced from 41GB to 19.9GB.

### System info (python version, jaxlib version, accelerator, etc.)

import jax
import flax.linen as nn
import numpy as np
from jax.experimental import mesh_utils
from jax.sharding import Mesh, NamedSharding, PartitionSpec
import os
from jax.experimental import pjit

emb_dim = 12288
vocab_size = 256000
batch_size = 1
max_seq_len = 8192

class Model(nn.Module):
    emb_dim: int
    vocab_size: int

    @nn.compact
    def __call__(self, x: jax.Array):
        layer = nn.Embed(num_embeddings=self.vocab_size, features=self.emb_dim, name="input_embed")
        x = layer(x)
        #x = nn.LayerNorm(use_bias=False, name="layernorm")(x)
        #x = layer.attend(x)
        return x

def gen_inputs():
    return np.random.randint(0, high=vocab_size - 1, size=(batch_size, max_seq_len))

devices = mesh_utils.create_hybrid_device_mesh((1,1), (1,1)) # (8,1) also works

MESH_DATA_AXIS = 'data'
MESH_TENSOR_AXIS = 'tensor'
mesh = Mesh(devices, axis_names=(MESH_DATA_AXIS, MESH_TENSOR_AXIS))

def P(*specs):
    return NamedSharding(mesh, PartitionSpec(*specs))

dummy_inputs = jax.jit(gen_inputs, out_shardings=P(MESH_DATA_AXIS, None))()
model = Model(emb_dim, vocab_size)

def get_shardings(param):
    if param.shape == (vocab_size, emb_dim):
        return P(None, MESH_TENSOR_AXIS)
    else:
        return P(None)

def init(inputs):
    key = jax.random.key(0)
    return model.init(key, inputs)

def forward(inputs, params):
    return model.apply(params, inputs)

abstract_vars = jax.eval_shape(init, dummy_inputs)
shardings = jax.tree_util.tree_map(lambda x: get_shardings(x), abstract_vars)
weights = jax.jit(init, out_shardings=shardings)(dummy_inputs)
for i in range(11):
    
    import ctypes
    libcudart = ctypes.cdll.LoadLibrary('libcudart.so')
    if i == 8 and (jax.process_index() == 0):
        libcudart.cudaProfilerStart()

    outputs = jax.jit(forward, out_shardings=P(MESH_DATA_AXIS, None, None))(dummy_inputs, weights)
    print(outputs.shape)

if (jax.process_index() == 0):
    libcudart.cudaProfilerStop()

**Comment by yashk2810** [comment]
Seems to be fixed by https://github.com/google/jax/commit/3f9540761e092772860cf7ca33a3cdca9ad40eb5. 

Please re-open this issue if it's not.

**Comment by froystig** [comment]
Yes, this should be covered by https://github.com/google/jax/commit/69878c4924f62a7b73d25c7721e5f07a75a96c18 and https://github.com/google/jax/commit/3f9540761e092772860cf7ca33a3cdca9ad40eb5 together. Thanks for noticing and closing.
